README.md
# ftdata
Unofficial CLI tool to retrieve device data and test results from the Focustronic API.

## Features
* List all devices within account to get device IDs
* Supports Alkatronic, Dosetronic, and Mastertronic
* Can output in table or CSV format
* Outputs to terminal or to file 

## Examples

1. Set credentials in your shell:
```shell
export FOCUSTRONIC_USERNAME='user here'
export FOCUSTRONIC_PASSWORD='password here'
```
or for Windows:
```shell
$env:FOCUSTRONIC_USERNAME = 'user here'
$env:FOCUSTRONIC_PASSWORD = 'password here'
```
2. List all the devices in your account:
```shell
$ ftdata list-devices
Device Name: Alkatronic, Device ID: 001
Device Name: Dosetronic, Device ID: 001
Device Name: Mastertronic, Device ID: 001
```
3. Pull Alkatronic records and display in the default table format:

```shell
$ ftdata get -t alkatronic -i 001
KhValue  Solution Added(mL)  Time
6.67     0.00                2023-02-03T20:09:44-06:00
6.67     0.00                2023-02-04T00:09:44-06:00
6.67     0.00                2023-02-04T04:09:34-06:00
6.67     0.00                2023-02-04T08:09:17-06:00
6.77     0.00                2023-02-04T12:09:28-06:00
6.67     0.00                2023-02-04T16:09:35-06:00
6.57     0.00                2023-02-04T20:09:30-06:00
```
4. Pull mastertronic records and output as CSV:
```shell
$ ftdata get -t mastertronic -i 001 -o csv
Parameter,Value,Time
ca,410.00,2023-02-05T09:33:22-06:00
no3,4.90,2023-02-05T10:23:52-06:00
oli,0.00,2023-02-07T23:03:51-06:00
no3,4.04,2023-02-08T00:21:51-06:00
ca,397.00,2023-02-08T05:24:41-06:00
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2023-03-04

- Initial release


cmd/utils_test.go
package cmd

import (
	"encoding/csv"
	"io"
	"os"
	"reflect"
	"testing"
)

// TestSliceInterfaceConverter tests the sliceInterfaceConverter function
func TestSliceInterfaceConverter(t *testing.T) {
	t.Parallel()

	// Test data
	testData := []string{"one", "two", "three"}

	// Expected result
	expected := []interface{}{"one", "two", "three"}

	// Actual result
	actual := sliceInterfaceConverter(testData)

	// Check if the actual result is equal to the expected result
	if len(actual) != len(expected) {
		t.Errorf("sliceInterfaceConverter() = %v, want %v", actual, expected)
	}
}

// // TestTimeFormatter tests the TimeFormatter function
// func TestTimeFormatter(t *testing.T) {
// 	t.Parallel()

// 	// Test data
// 	testData := int64(1675434543)

// 	// Expected result
// 	expected := "2023-02-03T08:29:03-06:00"

// 	// Actual result
// 	actual := TimeFormatter(testData)

// 	// Check if the actual result is equal to the expected result
// 	if actual != expected {
// 		t.Errorf("TimeFormatter() = %v, want %v", actual, expected)
// 	}
// }

// TestCsvWriter tests the csvWriter function
func TestCsvWriter(t *testing.T) {
	t.Parallel()

	// Test data
	headers := []string{"KhValue", "Solution Added", "Time"}
	rows := [][]string{
		{"7.00", "0.00", "2023-01-27T20:09:36-06:00"},
		{"7.01", "0.00", "2023-01-31T08:09:42-06:00"},
	}

	filename := "test.csv"
	defer func() {
		os.Remove(filename)
	}()

	// Call the function
	csvWriter(headers, rows, "test.csv")

	// Read the file
	file, err := os.Open(filename)
	if err != nil {
		t.Fatalf("failed to open file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	readHeaders, err := reader.Read()
	if err != nil {
		t.Fatalf("failed to read headers: %v", err)
	}
	if !reflect.DeepEqual(readHeaders, headers) {
		t.Fatalf("incorrect headers: expected %v, got %v", headers, readHeaders)
	}

	var readRows [][]string
	for {
		row, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatalf("failed to read row: %v", err)
		}
		readRows = append(readRows, row)
	}

	if !reflect.DeepEqual(readRows, rows) {
		t.Fatalf("incorrect rows: expected %v, got %v", rows, readRows)
	}

}


cmd/root.go
/*
Copyright © 2023 NAME HERE <EMAIL ADDRESS>
*/
package cmd

import (
	"io/ioutil"
	"log"
	"os"

	"github.com/spf13/cobra"
)

// rootCmd represents the base command when called without any subcommands
var rootCmd = &cobra.Command{
	Use:   "ftdata",
	Short: "ftdata is a CLI data retrieval tool for Focustronic product owners.",
	Long: `ftdata is used to retreive data for Focustronic product owners. It
can list all the devices in your account, and retreive data such as Alkatronic
and Mastertronic test results. It can also retrieve Dosetronic dosing records
as well. Examples:

ftdata list-devices
ftdata get --type alkatronic --id 001
ftdata get --type mastertronic --id 001 --output csv`,
	PersistentPreRun: func(cmd *cobra.Command, args []string) {
		if !Verbose {
			log.SetFlags(0)
			log.SetOutput(ioutil.Discard)
		}
	},
}

// Execute adds all child commands to the root command and sets flags appropriately.
// This is called by main.main(). It only needs to happen once to the rootCmd.
func Execute() {
	err := rootCmd.Execute()
	if err != nil {
		os.Exit(1)
	}
}

var Verbose bool

func init() {
	// Here you will define your flags and configuration settings.
	// Cobra supports persistent flags, which, if defined here,
	// will be global for your application.

	// rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.ftdata.yaml)")

	// Cobra also supports local flags, which will only run
	// when this action is called directly.
	//rootCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
	rootCmd.PersistentFlags().BoolVarP(&Verbose, "verbose", "v", false, "verbose level output")
	rootCmd.CompletionOptions.DisableDefaultCmd = true

}


cmd/utils.go
package cmd

import (
	"encoding/csv"
	"log"
	"os"
	"time"

	"github.com/fatih/color"
	"github.com/rodaine/table"
)

func sliceInterfaceConverter(s []string) []interface{} {
	i := make([]interface{}, len(s))
	for k, v := range s {
		i[k] = v
	}
	return i
}

func TimeFormatter(timeStamp int64) string {
	t := time.Unix(timeStamp, 0)
	f := t.Format(time.RFC3339)
	return f
}

// type TableStringer struct {
// 	table.Table
// }

// func (t TableStringer) String() string {
// 	return strings.TrimSpace(t.Table.String())
// }

func tableWriter(headers []string, rows [][]string) {
	headerFmt := color.New(color.FgGreen, color.Underline).SprintfFunc()
	columnFmt := color.New(color.FgYellow).SprintfFunc()

	h := sliceInterfaceConverter(headers)

	tbl := table.New(h...)
	tbl.WithHeaderFormatter(headerFmt).WithFirstColumnFormatter(columnFmt)

	for _, row := range rows {
		r := sliceInterfaceConverter(row)
		tbl.AddRow(r...)
	}

	tbl.Print()

}

func csvWriter(headers []string, rows [][]string, file string) {
	output := os.Stdout
	if file != "" {
		csvFile, err := os.Create(file)
		if err != nil {
			log.Fatalf("failed creating file: %s", err)
		}
		defer csvFile.Close()

		output = csvFile

		log.Printf("Saving CSV to %s", file)

	} else {
		log.Printf("Displaying CSV on stdout")
	}

	w := csv.NewWriter(output)
	defer w.Flush()

	if err := w.Write(headers); err != nil {
		log.Fatal("Error writing csv output: ", err)
	}

	for _, row := range rows {
		if err := w.Write(row); err != nil {
			log.Fatal("Error writing csv output: ", err)
		}
	}
}


cmd/version.go
/*
Copyright © 2023 NAME HERE <EMAIL ADDRESS>
*/
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"
)

const AppVersion = "1.0.0"

// versionCmd represents the version command
var versionCmd = &cobra.Command{
	Use:   "version",
	Short: "Print the version number of ftdata",
	Long:  "Print the version number of ftdata",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Printf("ftdata version: %s", AppVersion)
	},
}

func init() {
	rootCmd.AddCommand(versionCmd)

	// Here you will define your flags and configuration settings.

	// Cobra supports Persistent Flags which will work for this command
	// and all subcommands, e.g.:
	// versionCmd.PersistentFlags().String("foo", "", "A help for foo")

	// Cobra supports local flags which will only run when this command
	// is called directly, e.g.:
	// versionCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
}


cmd/listDevices.go
/*
Copyright © 2023 NAME HERE <EMAIL ADDRESS>
*/
package cmd

import (
	"fmt"
	"log"
	"os"

	"github.com/codyja/focustronic/api"
	"github.com/spf13/cobra"
)

// listDevicesCmd represents the listDevices command
var listDevicesCmd = &cobra.Command{
	Use:   "list-devices",
	Short: "List all the available devices in your account.",
	Long: `Lists all the available devices in you account
and prints the device ID.`,
	Run: func(cmd *cobra.Command, args []string) {

		username, ok := os.LookupEnv("FOCUSTRONIC_USERNAME")
		if !ok {
			log.Fatalf("FOCUSTRONIC_USERNAME not set")
		}
		password, ok := os.LookupEnv("FOCUSTRONIC_PASSWORD")
		if !ok {
			log.Fatalf("FOCUSTRONIC_PASSWORD not set")
		}
		client, err := api.NewFocustronicClient()
		if err != nil {
			fmt.Errorf("error initializing new Focustronic Client")
		}
		client.Authenticate(username, password)

		listDevices(client)
	},
}

func init() {
	rootCmd.AddCommand(listDevicesCmd)

	// Here you will define your flags and configuration settings.

	// Cobra supports Persistent Flags which will work for this command
	// and all subcommands, e.g.:
	// listDevicesCmd.PersistentFlags().String("foo", "", "A help for foo")

	// Cobra supports local flags which will only run when this command
	// is called directly, e.g.:
	// listDevicesCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
}

func listDevices(client *api.FocustronicClient) {
	allDevices, err := client.GetDevices()
	if err != nil {
		fmt.Println(err)
		log.Fatalf("Error getting devices: %v", err)
	}

	deviceList := client.ListDevices(allDevices)

	for k, v := range deviceList {
		fmt.Printf("Device Name: %s, Device ID: %v\n", v, k)
	}
}


cmd/get.go
/*
Copyright © 2023 NAME HERE <EMAIL ADDRESS>
*/
package cmd

import (
	"fmt"
	"log"
	"os"

	"github.com/codyja/focustronic/api"
	"github.com/spf13/cobra"
)

// getCmd represents the get command
var getCmd = &cobra.Command{
	Use:   "get",
	Short: "The get subcommand retreives records from the Focustronic API.",
	Long: `This get subcommand allows you to specify the device type (--type)
and device ID (--id) along with extra flags as needed. Example:

ftdata get -t mastertronic -i 001`,
	Run: func(cmd *cobra.Command, args []string) {

		username, ok := os.LookupEnv("FOCUSTRONIC_USERNAME")
		if !ok {
			log.Fatalf("FOCUSTRONIC_USERNAME not set")
		}
		password, ok := os.LookupEnv("FOCUSTRONIC_PASSWORD")
		if !ok {
			log.Fatalf("FOCUSTRONIC_PASSWORD not set")
		}
		client, err := api.NewFocustronicClient()
		if err != nil {
			fmt.Errorf("error initializing new Focustronic Client")
		}
		client.Authenticate(username, password)

		deviceType, _ := cmd.Flags().GetString("type")
		output, _ := cmd.Flags().GetString("output")
		days, _ := cmd.Flags().GetInt("days")
		deviceid, _ := cmd.Flags().GetInt("id")
		file, _ := cmd.Flags().GetString("file")
		GetRecords(client, deviceType, deviceid, days, output, file)

	},
}

var DeviceType, Output, File string
var Days, ID int

func init() {
	rootCmd.AddCommand(getCmd)

	// Here you will define your flags and configuration settings.

	// Cobra supports Persistent Flags which will work for this command
	// and all subcommands, e.g.:
	// getCmd.PersistentFlags().String("foo", "", "A help for foo")

	// Cobra supports local flags which will only run when this command
	// is called directly, e.g.:
	// getCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
	getCmd.Flags().StringVarP(&DeviceType, "type", "t", "", "Device type to retreive data for. Valid types are alkatronic, dosetronic, or mastertronic")
	getCmd.Flags().IntVarP(&ID, "id", "i", 0, "ID for device. Can be retrieved from the 'list-devices' command.")
	getCmd.Flags().IntVarP(&Days, "days", "d", 7, "Days worth of data to retrieve. Valid options are 7, 30, or 90.")
	getCmd.Flags().StringVarP(&Output, "output", "o", "table", "Output format. Valid options are table or csv.")
	getCmd.Flags().StringVarP(&File, "file", "f", "", "Output file name and location.")

}

func GetRecords(c *api.FocustronicClient, deviceType string, deviceId int, days int, output, file string) {

	if deviceType == "alkatronic" {
		records, err := c.GetAlkatronicRecords(deviceId, days)
		if err != nil {
			log.Fatalf("error getting alkatronic records: %v", err)
		}

		headers := []string{"KhValue", "Solution Added(mL)", "Time"}
		var rows [][]string

		for _, v := range records.Data {
			row := []string{
				fmt.Sprintf("%.2f", v.KhValue),
				fmt.Sprintf("%.2f", v.SolutionAdded),
				//time.Unix(v.CreateTime, 0).String(),
				TimeFormatter(v.CreateTime),
			}
			rows = append(rows, row)
		}

		if output == "table" {
			tableWriter(headers, rows)
		}
		if output == "csv" {
			csvWriter(headers, rows, file)
		}
	}

	if deviceType == "mastertronic" {
		records, err := c.GetMastertronicRecords(days, deviceId, "")
		if err != nil {
			log.Fatalf("Error getting record data: %v", err)
		}

		headers := []string{"Parameter", "Value", "Time"}
		var rows [][]string

		for _, v := range records.Data {
			row := []string{
				v.Parameter,
				fmt.Sprintf("%.2f", v.Value),
				//time.Unix(int64(v.RecordTime), 0).String(),
				TimeFormatter(v.RecordTime),
			}
			rows = append(rows, row)
		}

		if output == "table" {
			tableWriter(headers, rows)
		}
		if output == "csv" {
			csvWriter(headers, rows, file)
		}
	}

	if deviceType == "dosetronic" {
		records, err := c.GetDosetronicRecords(deviceId, days)
		if err != nil {
			log.Fatalf("Error getting record data: %v", err)
		}

		headers := []string{"PumpId", "DoseVolume", "DoseMode", "Time"}
		var rows [][]string

		for _, v := range records.Data {
			for _, v := range v {
				row := []string{
					fmt.Sprintf("%d", v.PumpID),
					fmt.Sprintf("%.2f", v.DoseVolume/100),
					fmt.Sprintf("%d", v.DoseMode),
					TimeFormatter(v.RecordTime),
				}

				rows = append(rows, row)

			}
		}

		if output == "table" {
			tableWriter(headers, rows)
		}
		if output == "csv" {
			csvWriter(headers, rows, file)
		}
	}
}
